### PR TITLE
Databox fixes

### PR DESCRIPTION
**Pull request by nilsdeppe**

### CODE REVIEW

File: src/Utilities/TMPL.hpp [hunk 1]
```diff
@@ -469,6 +469,29 @@ namespace tmpl = brigand;
 template <typename... Ts>
 using typelist = tmpl::list<Ts...>;
 
+namespace tmpl2 {
```
Comment (nilsdeppe) [inline src/Utilities/TMPL.hpp]: Document this into Utilities

File: src/DataStructures/DataBox.hpp [hunk 2]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+  static_cast<void>(std::initializer_list<char>{
+      (static_cast<void>(data.template get<Ts>() = make_deferred(helper(Ts{}))),
+       '0')...});
+}
+
+template <size_t ArgsIndex, typename Tag, typename... Tags, typename... Ts>
+SPECTRE_ALWAYS_INLINE constexpr bool add_item_to_box(
+    std::tuple<Ts...>& tupull,
+    detail::TaggedDeferredTuple<Tags...>&
+        data) noexcept(noexcept(data.template get<Tag>() =
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: check variables too

File: src/DataStructures/DataBox.hpp [hunk 3]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+      "Cannot pass a Deferred into the DataBox as an Item. This "
+      "functionally can trivially be added, however it is "
+      "intentionally omitted because users of DataBox are not "
+      "supposed to deal with Deferred.");
+  data.template get<Tag>() =
+      Deferred<item_type<Tag>>(std::move(std::get<ArgsIndex>(tupull)));
+  // If `tag` holds a Variables then add the contained Tensor's
+  add_variables_compute_tags_to_box<Tag>(
+      data, typename select_if_variables<Tag>::type{});
+  return true;
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: create a `cpp17::void_type` that can be changed to void when we have void objects

File: src/DataStructures/DataBox.hpp [hunk 4]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+    typename VariablesTag, typename... Ts, typename... Tags,
+    Requires<not tt::is_a_v<Variables, item_type<VariablesTag>>> = nullptr>
+SPECTRE_ALWAYS_INLINE constexpr void add_variables_compute_tags_to_box(
+    detail::TaggedDeferredTuple<Tags...>& /*data*/, typelist<Ts...> /*meta*/) {}
+
+template <typename VariablesTag, typename... Ts, typename... Tags,
+          Requires<tt::is_a_v<Variables, item_type<VariablesTag>>> = nullptr>
+SPECTRE_ALWAYS_INLINE constexpr void add_variables_compute_tags_to_box(
+    detail::TaggedDeferredTuple<Tags...>& data, typelist<Ts...> /*meta*/) {
+  auto helper = [lazy_function =
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: `const`

File: src/DataStructures/DataBox.hpp [hunk 5]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+template <typename FullTagLs, typename... Ts, typename... Tags,
+          typename... AddItemTags, typename... AddComputeItemTags, size_t... Is,
+          bool... DependenciesAddedBefore>
+SPECTRE_ALWAYS_INLINE void add_items_to_box(
+    std::tuple<Ts...>& tupull, detail::TaggedDeferredTuple<Tags...>& data,
+    tmpl::list<AddItemTags...> /*meta*/, std::index_sequence<Is...> /*meta*/,
+    tmpl::list<AddComputeItemTags...> /*meta*/) {
+  swallow(add_item_to_box<Is, AddItemTags>(tupull, data)...);
+  static_cast<void>(std::initializer_list<char>{
+      (add_compute_item_to_box<AddComputeItemTags, FullTagLs>(tupull, data),
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: don't need `tupull`

File: src/DataStructures/DataBox.hpp [hunk 6]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+      "supposed to deal with Deferred.");
+  data.template get<Tag>() =
+      Deferred<item_type<Tag>>(std::move(std::get<ArgsIndex>(tupull)));
+  // If `tag` holds a Variables then add the contained Tensor's
+  add_variables_compute_tags_to_box<Tag>(
+      data, typename select_if_variables<Tag>::type{});
+  return true;
+}
+
+template <typename ComputeItem, typename FullTagLs, typename... Tags,
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: Ls to List

File: src/DataStructures/DataBox.hpp [hunk 7]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+// clang-format off
+SPECTRE_ALWAYS_INLINE constexpr void add_compute_item_to_box(
+    std::tuple<Ts...>& /*tupull*/,
+    detail::TaggedDeferredTuple<Tags...>& data) noexcept(
+        noexcept(add_compute_item_to_box_impl<Tag, FullTagLs>(
+            data, typename Tag::argument_tags{}))) {
+  // clang-format on
+  add_compute_item_to_box_impl<Tag, FullTagLs>(data,
+                                               typename Tag::argument_tags{});
+  add_variables_compute_tags_to_box<Tag>(data,
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: use selec_if_variables to have only one function for add_compute_item

File: src/DataStructures/DataBox.hpp [hunk 8]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+    detail::TaggedDeferredTuple<Tags...>& data) {
+  add_reset_compute_item_to_box<Tag>(data, typename Tag::argument_tags{});
+  databox_detail::add_variables_compute_tags_to_box<Tag>(
+      data, typename item_type<Tag>::tags_list{});
+}
+
+template <typename DependencyGraph, typename current_vertex, typename... Tags>
+SPECTRE_ALWAYS_INLINE constexpr void reset_compute_items_apply(
+    detail::TaggedDeferredTuple<Tags...>& data) {
+  // using current_vertex = tmpl::front<Vertices>;
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: delete

File: src/DataStructures/DataBox.hpp [hunk 9]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+SPECTRE_ALWAYS_INLINE constexpr void reset_compute_item(
+    detail::TaggedDeferredTuple<Tags...>& data) {
+  add_reset_compute_item_to_box<Tag>(data, typename Tag::argument_tags{});
+}
+template <typename Tag, typename... Tags,
+          Requires<is_variables_compute_item<Tag>::value> = nullptr>
+SPECTRE_ALWAYS_INLINE constexpr void reset_compute_item(
+    detail::TaggedDeferredTuple<Tags...>& data) {
+  add_reset_compute_item_to_box<Tag>(data, typename Tag::argument_tags{});
+  databox_detail::add_variables_compute_tags_to_box<Tag>(
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: use selec_if_variables

File: src/DataStructures/DataBox.hpp [hunk 10]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+    tmpl::list<OldTags...> /*meta*/) {
+  static_cast<void>(std::initializer_list<char>{
+      (static_cast<void>(data.template get<OldTags>() =
+                             old_box.template get_lazy<OldTags>()),
+       '0')...});
+}
+
+template <typename DependencyGraph, typename... Tags,
+          typename... ComputeItemsToKeep>
+SPECTRE_ALWAYS_INLINE constexpr void reset_compute_items(
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: ??

File: src/DataStructures/DataBox.hpp [hunk 11]
```diff
@@ -321,11 +321,208 @@ class DataBox<TagsLs<Tags...>> {
...
+    detail::TaggedDeferredTuple<Tags...>& data,
+    tmpl::list<ComputeItemsToKeep...> /*meta*/);
+
+template <typename ComputeItem, typename... Tags, typename... ComputeItemTags>
+SPECTRE_ALWAYS_INLINE static constexpr void add_reset_compute_item_to_box(
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: try to just use compute item to box impl

File: src/DataStructures/DataBox.hpp [hunk 12]
```diff
@@ -337,19 +534,20 @@ constexpr DataBox<TagsLs<Tags...>>::DataBox(
...
-  detail::DataBoxAddHelper<full_tags_list, complete_tags_list>::
-      template add_items_to_box<0>(args_tuple, data_);
+  databox_detail::add_items_to_box<
+      typelist<FullItems..., FullComputeItems...>>(
+      args_tuple, data_, typelist<TagsInArgsOrder...>{},
+      std::make_index_sequence<sizeof...(TagsInArgsOrder)>{},
+      typelist<ComputeItemTags...>{});
 }
 
+
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: ??

File: src/DataStructures/DataBox.hpp [hunk 13]
```diff
@@ -711,13 +709,13 @@ using AddComputeItemsTags = tmpl::flatten<typelist<Tags...>>;
...
- * \tparam AddComputeItems list of \ref ComputeItemTag "compute item tags" to
- * add to the DataBox
+ * \tparam AddComputeItems list of \ref ComputeItemTag "compute item tags"
+ * to add to the DataBox
  *  \param args the data to be added to the DataBox
  */
 template <typename AddTags, typename AddComputeItems = typelist<>,
           typename... Args>
-SPECTRE_ALWAYS_INLINE constexpr auto create(Args&&... args) {
+SPECTRE_ALWAYS_INLINE constexpr auto create(Args&& ... args) {
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: ??

File: src/DataStructures/DataBox.hpp [hunk 14]
```diff
@@ -771,40 +771,36 @@ SPECTRE_ALWAYS_INLINE constexpr auto create_from(const Box& box,
...
-                "No items with the requested type were found in the DataBox");
-  ERROR("Cannot ever reach this function.");
+template <typename Tag, typename TagLs, typename T>
+constexpr void get_item_from_box_helper(const DataBox<TagLs>& box,
+                                        const std::string& tag_name,
+                                        T** result) {
+  if (get_tag_name<Tag>() == tag_name) {
+    // clang-tidy: do not use const_cast: in order to implement
+    // get_item_from_box non-recursively we need a const_cast
+    *result = const_cast<T*>(&box.template get<Tag>());  // NOLINT
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox.hpp]: See if const_cast can be avoided

File: docs/GroupDefs.hpp [hunk 15]
```diff
@@ -327,6 +343,7 @@
  * debugger, for example if you're using LLDB then run `lldb ./bin/RunTests`
  * and then to run the executable inside the debugger use `run Unit.Gradient`
  * inside the debugger.
+ *
```
Comment (nilsdeppe) [inline docs/GroupDefs.hpp]: ??

File: cmake/CompilationTestsParse.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+        compiler_and_message_regex = re.compile("\/\/ \[\[(COMPILER:.*?)\]\]",
+                                                re.DOTALL)
+        for compiler_and_message in re.findall(compiler_and_message_regex,
+                                               test_contents):
+            compiler_and_regex = re.search("COMPILER: (.*?)REGEX: (.*)",
+                                           compiler_and_message, re.DOTALL)
+
+            compilers = compiler_and_regex.group(1).strip().split(',')
+            regex_to_find = compiler_and_regex.group(2).replace("\n//", '')
+            if compilers[0] != "all":
```
Comment (nilsdeppe) [inline cmake/CompilationTestsParse.py]: user lower case comparison

File: cmake/SetupCompilationFailureTests.cmake [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+      set(ABSOLUTE_SOURCE_FILES "${ABSOLUTE_SOURCE_FILES};${SOURCE_FILE}")
+    endforeach()
+
+    execute_process(
+        COMMAND
+        ${PYTHON_EXECUTABLE}
+        ${CMAKE_SOURCE_DIR}/cmake/CompilationTestsParse.py
+        ${ABSOLUTE_SOURCE_FILES}
+        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/tmp)
+
```
Comment (nilsdeppe) [inline cmake/SetupCompilationFailureTests.cmake]: check return value of python script

File: src/Utilities/Overloader.hpp [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        F2(std::move(f2)),
+        Overloader<Fs...>(std::move(fs)...) {}
+
+  using F1::operator();
+  using F2::operator();
+  using Overloader<Fs...>::operator();
+};
+
+template <class F>
+struct Overloader<F> : F {
```
Comment (nilsdeppe) [inline src/Utilities/Overloader.hpp]: inherit privately

File: src/Utilities/TaggedTuple.hpp [hunk 19]
```diff
@@ -23,107 +260,595 @@
...
+        if (current_value < sizeof...(Tags)) {
+          os << ", ";
+        }
+      },
+      [&current_value, &os](
+          const auto& /*element*/,
+          const std::integral_constant<bool, false> /*meta*/) {
+        current_value++;
+        if (current_value < sizeof...(Tags)) {
+          os << ", ";
```
Comment (nilsdeppe) [inline src/Utilities/TaggedTuple.hpp]: do "not streamable"

File: docs/GroupDefs.hpp [hunk 20]
```diff
@@ -314,6 +314,29 @@
...
+ * since the file will not compile. The new file, say
+ * `Test_StaticAssertDataBox.cpp` must be added to the
+ * `SPECTRE_COMPILATION_TESTS` CMake variable, not `SPECTRE_TESTS`. Here is
+ * an example of how to write a compilation failure test:
+ *
+ * \snippet TestCompilationFramework.cpp compilation_test_example
+ *
+ * Each individual test must be inside an `#%ifdef COMPILATION_TEST_.*` block
+ * and each compilation test `cpp` file must contain
+ * `FILE_IS_COMPILATION_TEST` outside of any `#%ifdef`s.
```
Comment (nilsdeppe) [inline docs/GroupDefs.hpp]: "and it should be at the end of the file"

File: docs/GroupDefs.hpp [hunk 21]
```diff
@@ -314,6 +314,29 @@
...
+ *
+ * Specific compiler versions can be specified for which the regex changes.
+ * That is, the compiler version specified and all versions newer than that
+ * will use the regex, until a newer compiler version is specified. For
+ * example, the below checks for pre-GCC 5 and post-GCC 6.
```
Comment (nilsdeppe) [inline docs/GroupDefs.hpp]: pre-GCC 6

**Review by kidder — APPROVED** [review]

**Pull request merged**